fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java

```diff
@@ -2694,6 +2706,58 @@ public void addRebuildRecordCountsJob(List<CompletableFuture<Void>> work) {
         work.add(future);
     }
 
+    /**
+     * Return a list of boundaries separating the key ranges maintained by each FDB server. This information can be
+     * useful for splitting a large task (e.g., rebuilding an index for a large record store) into smaller tasks (e.g.,
+     * rebuilding the index for records in certain primary key ranges) more evenly so that they can be executed in a
+     * parallel fashion efficiently. The returned boundaries are an estimate from FDB's locality API and may not
+     * represent the exact boundary locations at any database version.
+     * <p>
+     * The boundaries are returned as a list which is sorted and does not contain any duplicates. The first element of
+     * the list is greater than or equal to <code>low</code>, and the last element is less than or equal to
+     * <code>high</code>.
+     * <p>
+     * This implementation may not work when there are too many shard boundaries to complete in a single transaction.
+     *
+     * @param low low endpoint of primary key range (inclusive)
+     * @param high high endpoint of primary key range (exclusive)
+     * @return the list of boundary primary keys
+     */
+    @API(API.Status.EXPERIMENTAL)
+    @Nonnull
+    public RecordCursor<Tuple> getPrimaryKeyBoundariesAsync(@Nonnull Tuple low, @Nonnull Tuple high) {
+        final Transaction transaction = ensureContextActive();
+        byte[] rangeStart = recordsSubspace().pack(low);
+        byte[] rangeEnd = recordsSubspace().pack(high);
+        CloseableAsyncIterator<byte[]> cursor = LocalityUtil.getBoundaryKeys(transaction, rangeStart, rangeEnd);
+        final boolean hasSplitRecordSuffix = hasSplitRecordSuffix();
+        DistinctFilterCursorClosure closure = new DistinctFilterCursorClosure();
+        return new IteratorCursor<>(getExecutor(), cursor)
+                .flatMapPipelined(
```

Comment:
Does this use `flatMapPipelined` instead of `mapPipelined` to handle the case where the cursor returns no elements?

Reply:
Yes. `mapPipelined` + `filter` also works but I think `flatMapPipelined ` is more clear

Reply:
I believe that because of the blocking nature of the `IteratorCursor` as the child cursor, you don't actually get any parallelism from `flatMapPipelined`. I think you would if the child cursors were `KeyValueCursor`s or if it used `mapPipelined`.